graylog2-server/src/main/java/org/graylog/plugins/views/search/Query.java

```diff
@@ -231,6 +250,14 @@ public Query addStreamsToFilter(Set<String> streamIds) {
         return toBuilder().filter(newFilter).build();
     }
 
+    public Query mapStreamCategoriesToIds(Function<Collection<String>, Stream<String>> categoryMappingFunction,
+                                          StreamPermissions streamPermissions) {
+        Set<String> streamIds = categoryMappingFunction.apply(usedStreamCategories())
+                .filter(streamPermissions::canReadStream)
+                .collect(toSet());
+        return addStreamsToFilter(streamIds);
+    }
```

Comment:
I am not sure if the solution is fully correct.
My doubts are:
1. You store stream categories in `filter` field, which represents a tree of filters, which could potentially be linked together with a complex OR/AND logic.
2. When fetching those stream categories, you seem to ignore where they are in this tree and treat them all as "top level" filters.

So, I'd say that:
1. If I've got smth wrong, just ignore this comment, please.
2. If stream categories are supposed to be connected with normal streams via OR/AND logic in a stream, you should probably revisit the mapping functionality and make it more complex...
3. If stream categories are meant to be a top-level filters only, maybe a new field for them would be a better and simpler way to handle things?

Reply:
`StreamCategoryFilters` are supposed to work basically exactly how `StreamFilters` work on queries, which are stored as filters in that map. The only difference is that the `StreamCategoryFilter` is mapped to a `StreamFilter` before search execution. The streams eventually just end up being added as part of the backend search query all on their own (at least from what I found in my testing), so that is what I did with the stream categories as well. So if there is more complex logic to be done, I'm not sure exactly what it would look like. @luk-kaminski

Reply:
With current structure, where you have:
```
    @Nullable
    @JsonProperty
    public abstract Filter filter();
```

`filter` could be an implementation of `OrFilter`, i.e.  containing 3 separate `StreamCategoryFilters`.
If I understand your code correctly, when doing the mapping you will use `addStreamsToFilter` method, effectively connecting those 3 filters via and logic, because:

```
    private Filter addStreamsTo(Filter filter, Set<String> streamIds) {
        final Filter streamIdFilter = StreamFilter.anyIdOf(streamIds.toArray(new String[]{}));
        if (filter == null) {
            return streamIdFilter;
        }
        return AndFilter.and(streamIdFilter, filter);
    }
```

So even with this simple scenario, where the tree of filters is not very deep, we have replaced the OR logic with AND logic, by accident.

AFAIK, we only use `addStreamsToFilter` method, when there are not streams at all, so in case of streams it probably does not produce errors.

That whole and/or filter logic may be an overkill, so I personally would use categories as a separate field to avoid even thinking about all the problems we may face with current solution. 
But maybe someone from architecture should have a say?

Reply:
This is the logic of that `StreamFilter.anyIdOf` that is being called in `addStreamsTo`.
```java
    public static Filter anyIdOf(String... ids) {
        final Set<Filter> streamFilters = Arrays.stream(ids)
                .map(StreamFilter::ofId)
                .collect(toSet());
        return OrFilter.builder()
                .filters(streamFilters)
                .build();
    }
```
So no, the `AndFilter` is being used with whatever the other filter provided is that is not the StreamFilter. It is being ANDed with the OrFilter of streams. An AndFilter of streams generally doesn't make much sense to me... messages can go to multiple streams, but that's typically not how we treat stream filters. They are always OrFilters.

Reply:
AndFilter of streams may not have sense, but it is possible for someone to create a search with that kind of filters and use it via the API... 
If you exclude a possibility of using some combinations of AndFilters/OrFilters and StreamCategoryFilters, shouldn't you have a validation of `filter` field of Query is constructed according to your expectations?

Would you be so kind to consider adding a unit test that shows how you imagine a change in the Query after invoking `addStreamsToQueriesWithCategories` for some of the filter scenarios? It may be useful anyway and it may help with the discussion.

Reply:
That's a fair point. I realize that just shoving the stream IDs to the top level of the filter can alter the original logic of the search query. I have reworked that code and instead of just adding all stream IDs to the top level we're now recursively iterating through the Filter tree and replacing in place any `StreamCategoryFilter` with its corresponding `StreamFilter`. This should allow for any more complex queries that may be generated through the API to maintain their original Filter logic. I also added some unit tests to verify the in place replacement works. Let me know what you think @luk-kaminski

Reply:
Thanks for that, @kingzacko1 !
It's been a heroic effort, considering a fact that universal character of the filters seems like an overkill in general...

I've left some new, minor comments, will mark this one as resolved.